### PR TITLE
replaceImage method as requested in #32

### PR DESCRIPTION
**Pull request by ccpandhare**
### What Changes this PR Contains
* Created the method `replaceImage`
* Small bugfix in `Run.js`
* Updated `Readme.md`
* Unit Test for `replaceImage`

### replaceImage Syntax
```js
sequencer.replaceImage('image_name','new_src');
```

### CODE REVIEW

**Comment by ccpandhare** [comment]
Should I merge this Is, @jywarren ?

**Comment by jywarren** [comment]
I think we need to say "imageId" instead and use getElementById, and why do we need newSrc? Shouldn't it just run the sequencer on the given image and replace it with the output?

Thanks!!!

**Comment by ccpandhare** [comment]
I am afraid not...
This method replaces the initial image in the sequencer, resulting in change in all images.
So we do need a new image for this, right?

**Comment by jywarren** [comment]
OK, so what i mean is, if you have an image on a webpage, what's everything you need to do in ImageSequencer to have a modified image replace that image? You need the `img` element's `id` so you can find it, and then you need some steps to run on it. The final state of hte image (after all the steps have run) should then be inserted back into the webpage, replacing the original. 

I notice, with chaining methods, (#40), you could do:

```js
$('#myImage').replace(sequencer.addStep(...).run($('#myImage')[0])); // but an example not in jQuery, maybe
```

**Comment by ccpandhare** [comment]
Okay. I thing the both of us have different understandings of what
`replaceImage` should do.

According to me:
Given an already present sequencer, the method changes the initial image,
thus, applying the same stpes to a completely different image and resulting
to a sequencer with initial image as taken from the user and steps
inherited from the previous image.

I believe you think otherwise. Is that so?

**Comment by ccpandhare** [comment]
Also, what you said will not be there for Node.js right?

**Comment by ccpandhare** [comment]
Now I do get what you were saying.

**Comment by jywarren** [comment]
Right, not for node -- and this is a slightly different use of IS -- the only state would be the steps, not the image. So we use it as if it were a filter for any given image. I guess in that usage, we might not want to actually store the image, or affect any images already loaded into the sequencer -- this is more just a useful way to use a sequencer that's already been set up, to do things on a webpage.

**Comment by ccpandhare** [comment]
So do you mean that a separate mode should be made in which the intermediate images aren't stored? Or maybe they're deleted shortly after? 
Actually, I am still not clear on your interpretation of `replaceImage`.

Is it a shorthand for `$('#myImage').replace(sequencer.addStep(...).run($('#myImage')[0]));` ?

Is it a shorthand for `sequencer.images[image].steps[0].draw(NEW_IMAGE)` ?

What exactly does it replace? (an image with an image in the DOM) or (an image with an image in the ImageSequencer) ?

**Comment by jywarren** [comment]
I think we could leave the intermediate images in place and if we never use the sequencer instance again, they'll just get abandoned, which for this use is fine. 

It is a shorthand for the first option -- the intention being to take an image from the DOM, run steps on it, and insert it back into the DOM where the original was. Imagine a use case -- you want to use IS to make every image in a web-based gallery monochrome. You could run (roughly):

```js
$('img.galleryImages').each(function(image) {
  var s = new ImageSequencer();
  s.addStep('monochrome').run(image, function onComplete(output) { $(image).replace(output) });
});
```

or in shortened syntax (and for just one image, in this case) something really minimal, like:

```js
ImageSequencer.replaceImage('img#myImage', steps);
```

I'm actually agnostic about whether we should use an existing instance of sequencer with steps already in it, or we should require steps to be added, and treat this as a kind of "quick constructor" -- that's up to you. I am mostly thinking about a convenient and compact **use case** and how to encourage people to easily use ImageSequencer in their web-based projects. 

Imagine wanting to make a quick JS utility to crop every image on a page, or generate thumbnails, or whatever -- what's the simplest and most straightforward way for IS to help people do this?

**Comment by ccpandhare** [comment]
Wow! That is a great insight, I must say!
Now this is clear to me. Thanks!

**Comment by jywarren** [comment]
Super, and sorry if I was mixing up the architectural question of state and so forth with the real desire which was just to increase adoption with an easy to use and short function! :-)

**Comment by ccpandhare** [comment]
No, what you said is perfectly fine!
Thanks!

**Comment by ccpandhare** [comment]
## Latest Changes
* I have created the `replaceImage` method.
* Added this update to README.md
* Created a small working demo at "replace-demo.html"

File: src/ReplaceImage.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      var raw = String.fromCharCode.apply(null,arr);
+      var base64 = btoa(raw);
+      var dataURL="data:image/"+ext+";base64," + base64;
+      make(dataURL);
+    };
+
+    if(url.substr(0,11).toLowerCase()!="data:image/") xmlHTTP.send();
+    else make(url);
+
+    function make(url) {
```
Comment (jywarren) [inline src/ReplaceImage.js]: What's the pros/cons of using this workaround here vs. just letting a canvas handle it?

**Comment by jywarren** [comment]
Cool, love the demo idea! Let's merge this as soon as we can as it's getting a bit complex and we really should be doing separate PRs for separate features.

File: README.md [hunk 2]
```diff
@@ -26,7 +26,21 @@ It is also for prototyping some other related ideas:
...
 
-## Usage
+## Quick Usage
+
+Image Sequencer can be used to run modules on an HTML Image Element using the
+`replaceImage` method. The method accepts two parameters - `selector` and `steps`.
+`selector` is a CSS selector. If it matches multiple images, all images will be
+modified. `steps` may be the name of a module or array of names of modules.
+
+Note: Local images will work only if they are in the same directory or a subdirectory.
```
Comment (jywarren) [inline README.md]: local images in the browser? why is this? is it relative path required? i.e. `../../` then we could say that. This is a bit mysterious sounding without an explanation.

File: README.md [hunk 3]
```diff
@@ -26,7 +26,21 @@ It is also for prototyping some other related ideas:
...
+
+```js
+  sequencer.replaceImage(selector,steps);
```
Comment (jywarren) [inline README.md]: can you show what `steps` should be?

File: replace-demo.html [hunk 4]
```diff
@@ -0,0 +1,63 @@
+<!DOCTYPE html>
```
Comment (jywarren) [inline replace-demo.html]: Can this go in the examples folder, as just `replace.html`?

**Comment by jywarren** [comment]
Also -- push a copy of this branch up to your gh-pages branch so we can try your demo! at https://ccpandhare.github.com/image-sequencer/examples/replace

File: README.md (hunk 2 above)
Comment (ccpandhare) [inline README.md]: Suppose the HTML file is in `/Users/PublicLab/Desktop/image-sequencer`, also:
image1 : ``
image2 : ``
image3: ``
image4: ``
image1, image2 are allowed.
image3, image4 will throw a CORS Error on some browsers like those running on Gecko.

File: README.md (hunk 3 above)
Comment (ccpandhare) [inline README.md]: Okay, I shall add that.

File: replace-demo.html (hunk 4 above)
Comment (ccpandhare) [inline replace-demo.html]: Yes, sorry.

File: README.md (hunk 2 above)
Comment (ccpandhare) [inline README.md]: I have updated README.md with a better explanation.

File: src/ReplaceImage.js (hunk 1 above)
Comment (ccpandhare) [inline src/ReplaceImage.js]: PROS:
* May bypass some CORS restrictions.
* Will also work on browsers not supporting Canvas. (read: IE < 9)

CONS:
* An extra HTTP request each time.

I feel this needs more testing (to find out whether the premise of bypassing CORS restrictions is actually valid). What do you say?

**Comment by ccpandhare** [comment]
Yes. I will add more intensive tests for `loadImage` and include tests for `replaceImage` once we merge this in.

**Comment by ccpandhare** [comment]
The demo is now live at [https://ccpandhare.github.io/image-sequencer/examples/replace.html](https://ccpandhare.github.io/image-sequencer/examples/replace.html) !

**Comment by ccpandhare** [comment]
Yes, options can be added. I can work on it. It is not exactly the same, as I haven't added support for options yet. It will hardly take a while. I will add that before merging. Is that okay?

Also this is my first demo for Image Sequencer! Do you like it? :-)

**Comment by jywarren** [comment]
Oh exciting! It works perfectly and looks great. Kudos!

**Comment by jywarren** [comment]
Aha, very interesting! If you repeat clicking too long, the image degrades; we can address this in a follow-up issue but it's really cool to see, actually! Must be decimal/integer issues. 

index 2

Yes go ahead and tweak the steps parameter as I think it's worth it for it to work the same way as addsteps, and you might even consider just passing it to addsteps to reduce code?

**Comment by ccpandhare** [comment]
> Oh exciting! It works perfectly and looks great. Kudos!

Thanks!

Wow! Interesting!
Actually this has to do with save-pixels.
save-pixels doesn't save images with 100% quality by default. That's why the image quality degrades with every use.

 > and you might even consider just passing it to addsteps to reduce code?

That's exactly whats happening. This is the code behind `replaceImages`:
```js
      this_.loadImage('default',url).addSteps('default',steps,options).run(function(out){
        the_image.src = out;
      });
```

**Comment by jywarren** [comment]
Oh perfect. Super!

**Comment by ccpandhare** [comment]
We can simply pass `{quality: 100}` to addSteps and that should fix this.
I did that before but for some weird reason my build fails when I do so (One unit test fails -- Double Inversion). That was a very weird issue, to be honest. Two seemingly identical stings were not identical! Anyways, that's for another time.

**Comment by jywarren** [comment]
Yes, mentioning that in the docs can be a succinct and clear way to show
proper usage. +1!

File: src/ReplaceImage.js (hunk 1 above)
Comment (jywarren) [inline src/ReplaceImage.js]: Sure. But I also like the nonrequirement of canvas, though of course some modules will decide to use it.

**Comment by ccpandhare** [comment]
Build Passed, merging it in!

**Comment by jywarren** [comment]
🎈

**Pull request merged**